src/internal/operators/delayWhen.ts

```diff
@@ -5,7 +5,10 @@ import { Subscription } from '../Subscription';
 import { OuterSubscriber } from '../OuterSubscriber';
 import { InnerSubscriber } from '../InnerSubscriber';
 import { subscribeToResult } from '../util/subscribeToResult';
-import { MonoTypeOperatorFunction, TeardownLogic } from '../types';
+import { MonoTypeOperatorFunction, TeardownLogic, ObservableInput } from '../types';
+
+/** @deprecated In future versions, empty notifiers will no longer re-emit the source value on the output observable. */
+export function delayWhen<T>(notifier: (value: T) => ObservableInput<never>);
```

Comment:
@benlesh I tried playing around with this a little, but I couldn't find a way to make all cases work (i.e., have deprecation show only if an empty observable is returned, but neither deprecation nor an error otherwise).

Reply:
@Airblader I've had a quick look at this. The basic problem is that the function didn't previously have overload signatures and - in adding one - you've hidden the function's signature, as when overload signatures are specified, the signature of the function's implementation is no longer public.

So you need to add two overload signatures, one deprecated and one not:

```ts
/* tslint:disable:max-line-length */
/** @deprecated In future versions, empty notifiers will no longer re-emit the source value on the output observable. */
export function delayWhen<T>(notifier: (value: T) => Observable<never>, subscriptionDelay?: Observable<any>);
export function delayWhen<T>(notifier: (value: T) => Observable<any>, subscriptionDelay?: Observable<any>);
/* tslint:disable:max-line-length */

/**
 * Delays the emission of items from the source Observable by a given time span
 * determined by the emissions of another Observable.
 * ...
 */
export function delayWhen<T>(delayDurationSelector: (value: T) => Observable<any>,
                             subscriptionDelay?: Observable<any>): MonoTypeOperatorFunction<T> {
  ...
```

Also, I would suggest leaving the return value typed as `Observable`. If the typings could/should be changed to accept functions returning `ObservableInput`, it would be preferable if that were done separately, perhaps when the test suite is better able to test types - which should be very soon.

Reply:
> So you need to add two overload signatures, one deprecated and one not:

I tried exactly this solution (with all the details you mentioned, except that yours are missing the return type), but 

```
/** @deprecated In future versions, empty notifiers will no longer re-emit the source value on the output observable. */
export function delayWhen<T>(notifier: (value: T) => Observable<never>, subscriptionDelay?: Observable<any>): MonoTypeOperatorFunction<T>;
export function delayWhen<T>(notifier: (value: T) => Observable<any>, subscriptionDelay?: Observable<any>): MonoTypeOperatorFunction<T>;
```

fails these tests for me:

```
// should be marked deprecated, but isn't
of(42).pipe(delayWhen(() => NEVER));
// should be marked deprecated, but isn't
of(42).pipe(delayWhen(value => NEVER));
// should not be marked deprecated, but is
const selector: (value: number) => Observable<any> = value => of(1337);
of(42).pipe(delayWhen(selector));
```

Reply:
Also, I think this deprecation notice is problematic for `NEVER` which might be a real usecase:

```
of(data).pipe(delayWhen(value => value >= 42 ? timer(1000) : NEVER))
```

Reply:
Similarly, I tried this as a standalone example and somehow all four calls are shown as deprecated:

```
interface Observable<T> {}

const ANY: Observable<any> = {} as Observable<any>;
const NEVER: Observable<never> = {} as Observable<never>;

/** @deprecated */
function delayWhen<T>(notifier: (value: T) => Observable<never>): void;
function delayWhen<T>(notifier: (value: T) => Observable<any>): void;

function delayWhen<T>(notifier: (value: T) => Observable<any>): void {
  return;
}

delayWhen<number>(() => NEVER);
delayWhen<number>(x => NEVER);
delayWhen<number>(() => ANY);
delayWhen<number>(x => ANY);
```

Reply:
Yeah, I missed the return type.

However, the overloads work fine for me, marking what should be deprecated and not marking what shouldn't.

I tested this by adding the first overload to the `delayWhen.d.ts` (within `node_modules/rxjs/...`) in a scratchpad project. What I added should be no different to what's emitted when the two overload declarations are included in the source.

I'm not sure how you are testing this, but the solution should work just fine. This how all of the other signature-specific deprecations work, so the mechanism should be sound.

Reply:
The idea of this deprecation notice, even if we get it to work, would by the way also mark `delayWhen(() => NEVER)` as deprecated even though it shouldn't as behavior won't change here. To be fair, though, having this in any application makes no sense.

Reply:
@cartant I don't doubt that something is fishy on my end here. Creating a new scratch file with

```
interface Observable<T> {}

const ANY: Observable<any> = {} as Observable<any>;
const NEVER: Observable<never> = {} as Observable<never>;

/** @deprecated */
function delayWhen<T>(notifier: (value: T) => Observable<never>): void;
function delayWhen<T>(notifier: (value: T) => Observable<any>): void;

function delayWhen<T>(notifier: (value: T) => Observable<any>): void {
  return;
}

delayWhen<number>(() => NEVER);
delayWhen<number>(x => NEVER);
delayWhen<number>(() => ANY);
delayWhen<number>(x => ANY);
```

shows all four calls as deprecated to me. I'll just update the PR as you suggested for now.